db/blob_file_state.h

```diff
@@ -0,0 +1,93 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+#include <iosfwd>
+#include <string>
+
+namespace rocksdb {
+
+constexpr uint64_t kInvalidBlobFileNumber = 0;
+
+class JSONWriter;
+class Slice;
+class Status;
+
+class BlobFileState {
+ public:
+  BlobFileState() = default;
+
+  BlobFileState(uint64_t blob_file_number, uint64_t total_blob_count,
+                uint64_t total_blob_bytes)
+      : blob_file_number_(blob_file_number),
+        total_blob_count_(total_blob_count),
+        total_blob_bytes_(total_blob_bytes) {}
+
+  BlobFileState(uint64_t blob_file_number, uint64_t total_blob_count,
+                uint64_t total_blob_bytes, uint64_t garbage_blob_count,
+                uint64_t garbage_blob_bytes)
+      : blob_file_number_(blob_file_number),
+        total_blob_count_(total_blob_count),
+        total_blob_bytes_(total_blob_bytes),
+        garbage_blob_count_(garbage_blob_count),
+        garbage_blob_bytes_(garbage_blob_bytes) {
+    assert(garbage_blob_count_ <= total_blob_count_);
+    assert(garbage_blob_bytes_ <= total_blob_bytes_);
+  }
+
+  uint64_t GetBlobFileNumber() const { return blob_file_number_; }
+
+  uint64_t GetTotalBlobCount() const { return total_blob_count_; }
+  uint64_t GetTotalBlobBytes() const { return total_blob_bytes_; }
+
+  void AddGarbageBlob(uint64_t size) {
+    assert(garbage_blob_count_ < total_blob_count_);
+    assert(garbage_blob_bytes_ + size <= total_blob_bytes_);
+
+    ++garbage_blob_count_;
+    garbage_blob_bytes_ += size;
+  }
+
+  uint64_t GetGarbageBlobCount() const { return garbage_blob_count_; }
+  uint64_t GetGarbageBlobBytes() const { return garbage_blob_bytes_; }
+
+  bool IsObsolete() const {
+    assert(garbage_blob_count_ <= total_blob_count_);
+
+    return !(garbage_blob_count_ < total_blob_count_);
+  }
+
+  void EncodeTo(std::string* output) const;
+  Status DecodeFrom(Slice* input);
+
+  std::string DebugString() const;
+  std::string DebugJSON() const;
+
+  bool operator==(const BlobFileState& rhs) const {
+    return blob_file_number_ == rhs.blob_file_number_ &&
+           total_blob_count_ == rhs.total_blob_count_ &&
+           total_blob_bytes_ == rhs.total_blob_bytes_ &&
+           garbage_blob_count_ == rhs.garbage_blob_count_ &&
+           garbage_blob_bytes_ == rhs.garbage_blob_bytes_;
+  }
+
+  bool operator!=(const BlobFileState& rhs) const { return !(*this == rhs); }
+
+ private:
+  uint64_t blob_file_number_ = kInvalidBlobFileNumber;
+  uint64_t total_blob_count_ = 0;
+  uint64_t total_blob_bytes_ = 0;
+  uint64_t garbage_blob_count_ = 0;
+  uint64_t garbage_blob_bytes_ = 0;
```

Comment:
Should we reserve a field of type std::string for full file checksum of blob files? cc @zhichao-cao

Reply:
Good point, thanks! Will reserve a field.

Reply:
@riversand963 That's a good point. I think yes, it will be good to add full file checksum for each blob file. @ltamasi Thanks for reserve that. I will discuss with you later and see how to add the checksum for blob files.

Reply:
@zhichao-cao Sounds good. Note that this is just groundwork; we can work on checksumming once blob files have been integrated into `Version`.

Reply:
@ltamasi Sure!